Add a `PermutationGroup.is_dihedral` property

skieffer opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234" (see
https://tinyurl.com/auto-closing for more information). Also, please
write a comment on that issue linking back to this pull request once it is
open. -->


#### Brief description of what is fixed or changed

Give the `PermutationGroup` class an `is_dihedral` property.

#### Other comments

Most of the other existing "named groups" (symmetric, alternating, cyclic) already have a corresponding `is_` property. However, while we do have a `DihedralGroup` function in `combinatorics/named_groups.py`, we don't currently have a matching `is_` property.

#### Release Notes

<!-- Write the release notes for this release below between the BEGIN and END
statements. The basic format is a bulleted list with the name of the subpackage
and the release note for this PR. For example:

* solvers
  * Added a new solver for logarithmic equations.

* functions
  * Fixed a bug with log of integers.

or if no release note(s) should be included use:

NO ENTRY

See https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more
information on how to write release notes. The bot will check your release
notes automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* combinatorics
  * Provide `PermutationGroup.is_dihedral` property
<!-- END RELEASE NOTES -->

skieffer committed:
Add a `PermutationGroup.is_dihedral` property

jksuom commented on sympy/combinatorics/perm_groups.py:
```diff
@@ -3230,6 +3231,103 @@ def is_cyclic(self):
+            if a < b:
+                x, y, a, b = y, x, b, a
+            # Using Theorem 2.1 of [3]:
+            if {a, b} == {2}:
```
The test could probably be `a == b == 2`. (No need to construct sets.)

skieffer replied:
Yeah, good call. This had crossed my mind, but I left it in there anyway.

jksuom commented on sympy/combinatorics/perm_groups.py:
```diff
@@ -3230,6 +3231,103 @@ def is_cyclic(self):
+                self._is_dihedral = True
+                return True
+            # Using Theorem 1.1 of [3]:
+            if (a, b) == (n, 2) and y*x*y == ~x:
```
I would consider writing `a == n and b == 2`.

skieffer committed:
Use local ref names

This prevents collisions with references defined under the `PermutationGroup`
class's docstring.

skieffer committed:
Use faster tests

smichr commented on sympy/combinatorics/perm_groups.py:
```diff
@@ -3230,6 +3231,103 @@ def is_cyclic(self):
+        order = self.order()
+
+        if order % 2 == 1:
+            self._is_dihedral = False
+            return False
```
If we are at Py 3.8 minimum, then is this a place where `return self._is_dihedral := False` could be used?

skieffer replied:
I'm entirely in favor of the walrus, and I went to go try this, but disappointingly it looks like it's not going to work.
I tried

    return (self._is_dihedral := False)

(the parentheses seem to be required, when it does work) but got

    SyntaxError: cannot use assignment expressions with attribute

So it seems it can't be used with attributes. Some explanation is offered at https://stackoverflow.com/a/60909096
Too bad.

smichr replied:
Thanks for checking and updating.

MarcoGorelli commented:
Hey, sorry for the intrusion, just noticed the comments:

> If we are at Py 3.8 minimum, then is this a place where return self._is_dihedral := False could be used?

> I'm entirely in favor of the walrus, and I went to go try this, but disappointingly it looks like it's not going to work.

Just FYI, there are many, many places in `sympy` where the walrus can be used, e.g.
```diff
--- a/sympy/combinatorics/prufer.py
+++ b/sympy/combinatorics/prufer.py
@@ -277,8 +277,7 @@ def edges(*runs):
             nmin = min(ei[0], nmin) if nmin is not None else ei[0]
             nmax = max(ei[1], nmax) if nmax is not None else ei[1]
             rv.append(list(ei))
-        missing = set(range(nmin, nmax + 1)) - got
-        if missing:
+        if (missing := set(range(nmin, nmax + 1)) - got):
             missing = [i + nmin for i in missing]
             if len(missing) == 1:
                 msg = 'Node %s is missing.' % missing.pop()
```

(found via [auto-walrus](https://github.com/MarcoGorelli/auto-walrus))
